### PR TITLE
Friendships v1

### PR DESCRIPTION
**Pull request by sinework**
### In this milestone:
Basic association  between friends is created

### The models have the following features:

- User can invite friends
- 
-  User can see friendships requests
- 
-  User has to accept a friendship request to start a friendship
- 
-  User can reject a friendship request

### CODE REVIEW

**Review by voscarmv — CHANGES_REQUESTED** [review]
# Nice work so far, team. 👍 

- Good work adding the necessary features on the user and friendship models.

## However, there are some changes required 👀 🧰 🔨 🔧 

- Your project is missing the following features which are required. Please make sure to add them before requesting a new code review:

- [x] User can invite friends
- [x] User can see friendships requests
- [x] User has to accept friendship request to start friendship
- [x] User can reject a friendship request
- [x] There is a button to invite user to friendship on users list page and on single user page
- [x] Posts timeline includes posts written by logged-in user and their friends
- [x] Please add some tests to your project!

- Also, please write a more detailed Pull Request description, where you summarize or list all the features you implemented.

***[OPTIONAL]***

- At this point you may want to start reading about the `n+1` problem in rails. Here's an article that you may use as a starting point: https://semaphoreci.com/blog/2017/08/09/faster-rails-eliminating-n-plus-one-queries.html

### Submit these changes and request a new code review! 😄 👍 

### Questions or comments? Reach me on [Twitter](https://twitter.com/voscarmv) or [LinkedIn](https://linkedin.com/in/oscar-mier) 🧑‍💻🗨️🧑‍💻 
# Happy Hacking! 🧠 ⌨️ 🖱️ 🖥️

**Review by acushlakoncept — CHANGES_REQUESTED** [review]
# Status: Changes Required ♻️

### Hello @sinework, great job implemented the user model queries and methods 👏 

Now that you've added those methods, let's implement them on the actual application using the controller. Currently, I do not seem to use those features requested when I logged in to your app. I will go ahead and list them below as well:

- [ ] When I sign in to the app I should be able to invite friends

- [ ] I should be able to see the friendships requests I have sent out

- [ ] I should also be able to accept or reject a friendship request to start or deny a friendship

- [ ] I should see a button to invite someone to friendship both on the users' list page and on the single-user page

- [ ] My Posts timeline should only include the posts written by me and that of those I have as friends.

- [x] I see you've done an excellent job on your unit test 👏 , kindly add integration tests as well.

Finally, I know you will have to write a couple of codes in your view files, ensure that you do not put logic in your views. The resources below could be of great help.

* [Where Do I Put My Code?](http://codefol.io/posts/Where-Do-I-Put-My-Code)
* [The Beginner's Guide to Rails Helpers](https://mixandgo.com/learn/the-beginners-guide-to-rails-helpers)
* [Rails View Helpers Best Practices](https://medium.com/@acushla4real/rails-view-helpers-best-practices-guide-d7d7b6db3db4)

### Cheers and Happy coding! 💻 👏

(Feel free to contact me on slack(@acushla), if you have any additional question)
Uduak Essien
[Twitter](https://twitter.com/acushlakoncept) | [LinkedIn](https://www.linkedin.com/in/acushlakoncept/)

**Review by teekaytech — CHANGES_REQUESTED** [review]
## Status: Changes Required :warning: :vertical_traffic_light: 

### Hello team :wave:

__You've made laudable progress in this milestone. I particularly appreciate the way you completely moved your logic out of the views.
However, some changes are required before this project can be approved. Please check my comments below and fix them as appropriate.__

**[ REQUIRED CHANGES ]**

- [x] Your `.ruby-version` file stated that ruby v2.7.0 is used, while your Gemfile used v2.7.1 You should be consistent with the ruby version across all files. Kindly stick to a single version.

- [x] You're having view files for friendship actions. Kindly delete these views as they're not used in your application.

- [x] You should also remove the username and password settings that you included in your `database.yml`. This might work for you, but your username and password cannot work for everyone. You can set it up for your development, but not as the default setting.

- [x] Some of your newly added integration test suites are failing. Kindly check and fix them as appropriate.



* * *
**Please make the required changes and submit another review request when you are done.**

* * *

**Taofeek Olalere
[Twitter](https://twitter.com/ola_lere) | [Github](https://github.com/teekaytech) | @taofeekolalere on Slack**
__Happy Coding :smiley:__

File: config/database.yml
Comment (elmejdki) [inline config/database.yml]: remove these two lines or at least mention in the readme file that the users should use their corresponding username and password.

**Review by elmejdki — CHANGES_REQUESTED** [review]
# status: require changes :recycle: 

Great job so far @team :clap: :clap: :clap:, but you still have some change :hammer: :hammer_and_wrench: :wrench: :toolbox: to take care of in order to get this project approved :checkered_flag:.

## Required Changes:

- [ ] Your .ruby-version file stated that ruby v2.7.0 is used, while your Gemfile used v2.7.1 You should be consistent with the ruby version across all files. Kindly stick to a single version.

image

- [ ] as mentioned by the previous TSE, your tests are falling, please let us know if there are working on your end, to see how to fix them.

### Please make all the required changes before you submit another code review.
## Happy Coding :tophat: 
_Please don't hesitate to contact me on slack if you have any question_

**Comment by sinework** [comment]
Hello there, I have removed the DB credentials. However, as you can see below:
### 1 My ruby version and the one stated in the gem file are both version 2.7.1, hence, I'm not running into any conflicts on my PC
ruby-v
GemLock
Can you kindly check which version of ruby your machine is running?
2. No test case is failing on my machine. 
NoTestFailing

**Review by andresporras3423 — CHANGES_REQUESTED** [review]
### Status: Changes Required :recycle:
**Hello Sine Shiferaw** 
Good job with the improvements for this project :muscle:  :muscle:
It meets most of the requirements :clap: :clap: :clap:  
There are just a few details to correct before you can go ahead:

- [x] Remove commented line of code in post_controller.rb
- [x] Friends' posts appear repeated in "timeline"
- [x] Your project failed one of your tests
 

 
Please request another review once you are ready! :+1: 
 
Happy coding :relaxed:
 
Oscar Andrés Russi Porras

**Review by lekegitrepo — CHANGES_REQUESTED** [review]
## Status: Changes Required! :recycle:
### Hi @sinework :clap: :clap:
#### Good job so far :+1: However, the app is currently not working. Check images below to see exactly what went wrong.

#### Tests:
- _Four integration tests are failing_
  
Screenshot from 2020-12-15 10-28-07
Screenshot from 2020-12-15 10-27-30

#### Application Behaviour:
- _It outputted an error when I tried to load the page_

Action-Controller-Exception-caught

#### _Reviewed by Temitayo Adeleke_

**Review by javitocor — CHANGES_REQUESTED** [review]
# STATUS CHANGES REQUIRED 🛑 

A great job implementing the suggested changes by previous TSE 👍 
But we still need to do something else in order to get the project approved 💪 
Kindly check my comments below 👇  and resubmit your project once done ⏩ 
Good luck! 😃 
## Mandatory changes
- The integration tests are failing, kindly fix them

**Comment by sinework** [comment]
I have been testing and testing and retesting the project the whole day. I have had four microverse students run the test and not a single test is failing on the machines it's checked in. I've reported that I haven't been able to see the failing test on my side on one of the previous reviews. Please kindly tell me which exact tests are failing and how exactly I can determine what's failing when I don't see the failures on my machine. 
NoTestFailing

**Review by teekaytech — APPROVED** [review]
# Status: Approved :green_circle: 

### Congratulations team, this project has passed all the mandatory requirements. It is therefore approved :heavy_check_mark:


You can merge to master.
Good luck with your next projects.

* * * 
**Taofeek Olalere
[Twitter](https://twitter.com/ola_lere) | [Github](https://github.com/teekaytech) | @taofeekolalere on Slack**
__Happy Coding :smiley:__

**Comment by sinework** [comment]
Thank you!

**Pull request merged**
